File: src/main/java/com/ifat/bdd/spark/labs/lab_taxi_lab/mainly_java_form/service/TaxiService.java
package com.ifat.bdd.spark.labs.lab_taxi_lab.mainly_java_form.service;

import com.ifat.bdd.spark.labs.lab_taxi_lab.mainly_java_form.converter.DriverUtil;
import com.ifat.bdd.spark.labs.lab_taxi_lab.mainly_java_form.converter.TripsInfoUtil;
import com.ifat.bdd.spark.labs.lab_taxi_lab.mainly_java_form.model.Trip;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

import java.time.LocalDate;
import java.util.List;
import java.util.logging.Logger;

public class TaxiService {
    private final String applicationName = "Taxi App";
    private SparkConf sparkConf;
    private JavaSparkContext sc;
    private JavaRDD<String> tripsRdd;
    private JavaRDD<String> driversRdd;

    public TaxiService(String tripsInputFile, String driversInputFile) {
        sparkConf = new SparkConf().setAppName(applicationName).setMaster("local[*]");
        sc = new JavaSparkContext(sparkConf);
        tripsRdd = sc.textFile(tripsInputFile);
        driversRdd = sc.textFile(driversInputFile);
    }

    public long findNumberOfLines() {
        return tripsRdd.count();
    }

    public long findNumberOfTripsSameCityLongerThenX(final String city, final int length) {
        return tripsRdd.map(TripsInfoUtil::ConvertToTrips)
                .filter(t -> t.getTripTargetCity().equalsIgnoreCase(city) && t.getTripLength() >= length)
                .count();
    }

    public Double totalSumOfTripsToSameCity(final String city) {
        return tripsRdd.map(TripsInfoUtil::ConvertToTrips)
                .filter(t -> t.getTripTargetCity().equalsIgnoreCase(city))
                .mapToDouble(Trip::getTripLength)
                .sum();
    }

    public List<String> top3WithMaxTripLengthOnGivenDay(LocalDate date) {
        JavaPairRDD<String, String> drivers = driversRdd.map(DriverUtil::DriversInfoConverter)
                .mapToPair(t -> new Tuple2<>(t.getDriverId(), t.getDriverName()));

        return
                tripsRdd.map(TripsInfoUtil::ConvertToTrips)
                        .filter(t -> t.getTripDateTime().toLocalDate().isEqual(date))
                        .mapToPair(t -> new Tuple2<>(t.getDriverId(), t.getTripLength()))
                        .reduceByKey(Integer::sum)
                        .join(drivers)
                        .values()
                        .sortBy(integerStringTuple2 -> integerStringTuple2._1, false, 1)
                        .map(p -> p._2)
                        .take(3);

    }


}

File: src/main/java/com/ifat/bdd/spark/labs/lab_taxi_lab/mainly_scala_form/service/TaxiService.scala
package com.ifat.bdd.spark.labs.lab_taxi_lab.mainly_scala_form.service

import java.time.LocalDate

import com.ifat.bdd.spark.labs.lab_taxi_lab.mainly_scala_form.converter.Conversions.{DriversExtension, TripExtension}
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

class TaxiService(tripsInputFile: String, driversInputFile: String) {
  private val sparkConf = new SparkConf().setAppName("Trip App").setMaster("local[*]")
  private val sc = new SparkContext(sparkConf)
  private val tripsRdd: RDD[String] = sc.textFile(tripsInputFile)
  private val driversRdd: RDD[String] = sc.textFile(driversInputFile)

  def findNumberOfLines():Long = tripsRdd.count()

  def findNumberOfTripsSameCityLongerThenX(city: String, length: Int): Long = {
    tripsRdd.map(l => l.convertToTrip())
      .filter(t => t.tripTargetCity.equalsIgnoreCase(city) && t.tripLength >= length)
      .count()
  }

  def totalSumOfTripsToSameCity(city: String): Double = {
    tripsRdd.map(l => l.convertToTrip())
      .filter(t => t.tripTargetCity.equalsIgnoreCase(city))
      .map(t => t.tripLength)
      .sum()
  }


  def top3WithMaxTripLengthOnGivenDay(date: LocalDate): Array[String] = {
    val drivers: RDD[(String, String)]
    = driversRdd.map(l => l.convertToDriver()).map(d => Tuple2(d.driverId, d.driverName))

    tripsRdd.map(l => l.convertToTrip())
      .filter(t => t.tripDateTime.isEqual(date))
      .map(t =>  Tuple2(t.driverId, t.tripLength))
      .reduceByKey((l1, l2) => l1 + l2)
      .join(drivers)
      .values
      .sortBy(t => t._1, false:Boolean, 1)
      .map(p => p._2)
      .take(3)

  }
}


